README.md
# JBookTrader [![Build Status](https://travis-ci.org/jeremyis78/JBookTrader.svg?branch=master)](https://travis-ci.org/jeremyis78/JBookTrader)

JBookTrader is a fully automated trading system (ATS) that can trade various
types of market securities during the trading day without user monitoring. All
aspects of trading, such as obtaining market prices, analyzing price patterns,
making trading decisions, placing orders, monitoring order executions, and 
controlling the risk are automated according to the user preferences. The 
central idea behind JBookTrader is to completely remove emotions from trading, 
so that the trading system can systematically and consistently follow a 
predefined set of rules.

The features include strategy back testing, optimization, market data recording,
and real time trading via the Interactive Brokers API.

JBookTrader is written in Java and is intended for software developers. It is
not an "off-the-shelf" product that can be installed and run. Instead,
JBookTrader provides a framework for developing automated trading systems and
requires a certain amount of programming knowledge and experience in Java. If
you are not a software developer or if you don't have much experience
programming in Java, JBookTrader is probably not for you.

Before you start working with JBookTrader, please read the 
[JBookTrader User Guide](
https://github.com/jeremyis78/JBookTrader/tree/master/docs/UserGuide) 
and the documents located in the /docs directory of the distribution.

![Back Testing Screenshot](docs/backtest.png "Back Testing")
![Charting Screenshot](docs/chart.png "Charting")
![Optimizing Visualization](docs/optMap2.png "Optimizing Visualization")


### Development
Once you've read the user guide and docs, you can get started.
JBookTrader can be built with Java 8. 
And in fact, the travis CI build uses jdk8. 
  
Development cycle
NOTE: https://mvnrepository.com/artifact/com.interactivebrokers.tws/tws-api/9.72.05 
NOTE: The latest [tws api version (9.72.05) fails to compile](
https://github.com/jeremyis78/JBookTrader/issues/2)

```bash
mvn clean compile
```
```bash
./run/JBookTrader.sh
```  
  
Packaged jar:
```bash
mvn clean package
```  
```bash
./run/JBookTraderJar.sh
```

Release [still a work in progress](https://github.com/jeremyis78/JBookTrader/issues/4)

```bash
BUILD_NUMBER=...
mvn deploy scm:tag -Drevision=${BUILD_NUMBER}
```

CHANGELOG.md
* Include decompiled source for tws-api v9.66
* Use mvn to build project

9.01 (July 8, 2013)
-------------------

* Email/SMS notifications
* Third party library updated: IB API from 9.63 to 9.66, JFreeChart from 1.0.13 to 1.0.14, JCommon from 1.0.16 to 1.0.17, JCalendar from 1.3.3 to 1.4
* Indicator initialization improved
* Fixed compilation time warnings related to generics
* Added a build script for building JBookTrader in an OS-independent and IDE-independent way
* Backtester and optimizer cache the data sets for fast loading
* Ability to run JBookTrader from a single JAR
* Ability to switch to "forward test" mode and to suspend live trading
* New performance metric: CPI
* Portfolio Manager for dynamic position sizing
* Fixed failing unit tests
* Rejected orders are handled correctly
* Ability to trade an FA account
* Calculation of "most liquid contract" for the CL
* Very efficient calculation if min/max in a moving window, as well as standard deviation
* Ability to force-close the positions if the connectivity gap is wide enough
* Exchange hours detection
* Fixed thread safety issues in optimizer
* Fixed average duration calculation
* Prevent entering the position when less than 15 minutes left in the trading interval
* Rearranged the packages
* Max single loss calculations


8.07 (January 9, 2012)
----------------------

* Improved the methodology of indicator initialization
* Fixed date scroll issues in the charts
* Optimization results are now sortable by table column
* Fixed the problem associated with the 2104 error code


8.06 (April 24, 2011)
---------------------

* Strategy information dialog uses streaming book information
* Trades are annotated slightly differently on the performance chart


8.05 (March 26, 2011)
---------------------

* Fixed the "cancel" functionality for the optimization and backtesting dialogs
* Removed the C2 feature
* Fixed a problem preventing two instances of the same indicator being shown on the chart
* Added an "inclusion" criteria for the optimization dialog


8.04 (February 27, 2011)
------------------------

* Revised Tension indicator
* Changed North America bundled future commissions from $2.00 to $2.01
* Strategy report is now more readable
* Revised sample strategies


8.03 (February 12, 2011)
------------------------

* Refactored various pieces for clarity and performance


8.02 (November 24, 2010)
------------------------

* Both the "brute force" and the "divide-and-conquer" optimizers run about 2 times faster


8.01 (November 14, 2010)
------------------------

* Added support for volume
* Added new indicators
* Fixed and refactored MovingWindow? classes
* Added a new "bias" metric


7.10 (October 29, 2010)
-----------------------

* Added 3 new indicators, including Kalman filter
* Added the "settlement time" for better backtesting
* Both "brute force" and "divide-and-conquer" optimizers run about 1.3 to 2.0 times faster
* Improved the MovingWindow? class
* Fixed a minor problem with a unicode character


7.09 (October 11, 2010)
-----------------------

* Refactored several classes
* Fixed NTP clock timeouts and shifts
* Added the "Average Duration" performance metric


7.08 (October 6, 2010)
----------------------

* Fixed the issue of leaving open positions overnight
* Fixed US futures commissions
* Fixed market depth resetting issues
* Improved D&C optimizer
* Improved web interface


7.07 (September 8, 2010)
------------------------

* Set look and feel to Nimbus and removed other LAFs
* Simplified web interface
* Reorganized indicators
* Simplified backtesting dialog
* Improved reporting
* Simplified performance charts
* Fixed a recursive call issue with NTP clock
* Fixed the progress indicator issue in divide-and-conquer dialog
* Reorganized sample strategies


7.06 (June 12, 2010)
--------------------

* Upgraded from Substance 5.2 to 6.0
* Added three more types of look and feel: Liquid, SeaGlass?, and Nimbus
* Simplified Windows start-up script
* Fixed memory leak in back tester
* Added two more sample strategies


7.05 (May 31, 2010)
-------------------

* Added new indicators
* Improved performance of certain indicators, such as Bollinger, PriceSMA, and Price Volatility
* Added a bar size option to the back tester dialog
* Upgraded IB API to version 9.63
* Added new strategies
* Added new data set for back testing


7.04 (September 6, 2009)
------------------------

* Fixed an NTP clock issue which caused "time shifts"
* Reports use NTP clock
* Reports format and appearance revised
* IB error 322 is handled


7.03 (August 24, 2009)
----------------------

* 1-second market depth representation changed
* JBT is driven by atomic clock
* Added about 15 indicators
* Collective2 code restructured
* Fixed some issues in web admin console
* Optimizers are faster
* Fixed a concurrency issue in optimization map
* Fixed order processing issue which caused the "Order would cross related resting order" message and order cancellation
* Added sample strategies


7.02 (July 28, 2009)
--------------------

* Revised and improved the "divide and conquer" optimizer
* Both "brute force" and "divide and conquer" optimizers are faster
* Web console is more dynamic
* Range of dates can be specified in the backtest and optimization dialogs
* Fixed scroll bar and zooming issues in performance chart
* Added "user manual" and "release notes" to main menu
* Market depth validation is stricter and more consistent
* Better look and feel with the substance L&F
* User manual updated


7.01 (July 13, 2009)
--------------------

* Fixed the "market depth reset" bug which caused the market data "freezing"
* Replaced the "liquid" L&F with the "substance" L&F
* Functionality and appearance of web console significantly improved
* Updated user manual
* Added a new manual, describing how to set up JBT in Eclipse on Mac OS
* Minor GUI tweaks for Mac OS
* Optimizer window "remembers" its last size and position
* Updated sample strategies



tws-api/src/main/java/com/ib/client/TickType.java
//
// Source code recreated from a .class file by IntelliJ IDEA
// (powered by Fernflower decompiler)
//

package com.ib.client;

public class TickType {
    public static final int BID_SIZE = 0;
    public static final int BID = 1;
    public static final int ASK = 2;
    public static final int ASK_SIZE = 3;
    public static final int LAST = 4;
    public static final int LAST_SIZE = 5;
    public static final int HIGH = 6;
    public static final int LOW = 7;
    public static final int VOLUME = 8;
    public static final int CLOSE = 9;
    public static final int BID_OPTION = 10;
    public static final int ASK_OPTION = 11;
    public static final int LAST_OPTION = 12;
    public static final int MODEL_OPTION = 13;
    public static final int OPEN = 14;
    public static final int LOW_13_WEEK = 15;
    public static final int HIGH_13_WEEK = 16;
    public static final int LOW_26_WEEK = 17;
    public static final int HIGH_26_WEEK = 18;
    public static final int LOW_52_WEEK = 19;
    public static final int HIGH_52_WEEK = 20;
    public static final int AVG_VOLUME = 21;
    public static final int OPEN_INTEREST = 22;
    public static final int OPTION_HISTORICAL_VOL = 23;
    public static final int OPTION_IMPLIED_VOL = 24;
    public static final int OPTION_BID_EXCH = 25;
    public static final int OPTION_ASK_EXCH = 26;
    public static final int OPTION_CALL_OPEN_INTEREST = 27;
    public static final int OPTION_PUT_OPEN_INTEREST = 28;
    public static final int OPTION_CALL_VOLUME = 29;
    public static final int OPTION_PUT_VOLUME = 30;
    public static final int INDEX_FUTURE_PREMIUM = 31;
    public static final int BID_EXCH = 32;
    public static final int ASK_EXCH = 33;
    public static final int AUCTION_VOLUME = 34;
    public static final int AUCTION_PRICE = 35;
    public static final int AUCTION_IMBALANCE = 36;
    public static final int MARK_PRICE = 37;
    public static final int BID_EFP_COMPUTATION = 38;
    public static final int ASK_EFP_COMPUTATION = 39;
    public static final int LAST_EFP_COMPUTATION = 40;
    public static final int OPEN_EFP_COMPUTATION = 41;
    public static final int HIGH_EFP_COMPUTATION = 42;
    public static final int LOW_EFP_COMPUTATION = 43;
    public static final int CLOSE_EFP_COMPUTATION = 44;
    public static final int LAST_TIMESTAMP = 45;
    public static final int SHORTABLE = 46;
    public static final int FUNDAMENTAL_RATIOS = 47;
    public static final int RT_VOLUME = 48;
    public static final int HALTED = 49;
    public static final int BID_YIELD = 50;
    public static final int ASK_YIELD = 51;
    public static final int LAST_YIELD = 52;
    public static final int CUST_OPTION_COMPUTATION = 53;
    public static final int TRADE_COUNT = 54;
    public static final int TRADE_RATE = 55;
    public static final int VOLUME_RATE = 56;
    public static final int LAST_RTH_TRADE = 57;

    public TickType() {
    }

    public static String getField(int tickType) {
        switch(tickType) {
            case 0:
                return "bidSize";
            case 1:
                return "bidPrice";
            case 2:
                return "askPrice";
            case 3:
                return "askSize";
            case 4:
                return "lastPrice";
            case 5:
                return "lastSize";
            case 6:
                return "high";
            case 7:
                return "low";
            case 8:
                return "volume";
            case 9:
                return "close";
            case 10:
                return "bidOptComp";
            case 11:
                return "askOptComp";
            case 12:
                return "lastOptComp";
            case 13:
                return "modelOptComp";
            case 14:
                return "open";
            case 15:
                return "13WeekLow";
            case 16:
                return "13WeekHigh";
            case 17:
                return "26WeekLow";
            case 18:
                return "26WeekHigh";
            case 19:
                return "52WeekLow";
            case 20:
                return "52WeekHigh";
            case 21:
                return "AvgVolume";
            case 22:
                return "OpenInterest";
            case 23:
                return "OptionHistoricalVolatility";
            case 24:
                return "OptionImpliedVolatility";
            case 25:
                return "OptionBidExchStr";
            case 26:
                return "OptionAskExchStr";
            case 27:
                return "OptionCallOpenInterest";
            case 28:
                return "OptionPutOpenInterest";
            case 29:
                return "OptionCallVolume";
            case 30:
                return "OptionPutVolume";
            case 31:
                return "IndexFuturePremium";
            case 32:
                return "bidExch";
            case 33:
                return "askExch";
            case 34:
                return "auctionVolume";
            case 35:
                return "auctionPrice";
            case 36:
                return "auctionImbalance";
            case 37:
                return "markPrice";
            case 38:
                return "bidEFP";
            case 39:
                return "askEFP";
            case 40:
                return "lastEFP";
            case 41:
                return "openEFP";
            case 42:
                return "highEFP";
            case 43:
                return "lowEFP";
            case 44:
                return "closeEFP";
            case 45:
                return "lastTimestamp";
            case 46:
                return "shortable";
            case 47:
                return "fundamentals";
            case 48:
                return "RTVolume";
            case 49:
                return "halted";
            case 50:
                return "bidYield";
            case 51:
                return "askYield";
            case 52:
                return "lastYield";
            case 53:
                return "custOptComp";
            case 54:
                return "trades";
            case 55:
                return "trades/min";
            case 56:
                return "volume/min";
            case 57:
                return "lastRTHTrade";
            default:
                return "unknown";
        }
    }
}
